Adds Migrate from VS Code to Positron tutorial (#415)
posit-dev/positron-docs: tutorial-migrate-from-vscode.qmd

---
title: "Migrate from VS Code to Positron"
description: "Import your VS Code settings and extensions into Positron and see what makes Positron different."
---

::: {.callout-note}
Is this the right tutorial to start with? This tutorial picks up where [First data analysis with Python in a Jupyter Notebook](tutorial-get-started-ipynb.qmd) leaves off, using the same project. If you have not done that tutorial yet, start there. It takes about 15 minutes and leaves you with a project you can configure here.
:::

You just finished your first analysis in Positron. If a few things felt slightly unfamiliar compared to VS Code, this tutorial explains them, and shows you why they work the way they do.

> Good news: as a VS Code user, you already know most of Positron. Positron builds on [Code OSS](https://github.com/microsoft/vscode), the open source core of VS Code. These parts of Positron all work the way you expect, so you do not need to relearn them: 

Reviewer: ```suggestion
The good news is as a VS Code user, you already know most of Positron. Positron builds on [Code OSS](https://github.com/microsoft/vscode), the open source core of VS Code. These parts of Positron all work the way you expect, so you do not need to relearn them: 
```

- Editor
- Command Palette
- Terminal
- Git client
- Extensions system

What is different is everything Positron adds for data science: 

- a first-class Console
- a live Variables tracker
- a Plots pane
- a Data Explorer
- native Python and R support with no setup 

Over the next 20 minutes you will bring your VS Code setup to Positron, then meet the tools that make Positron more than a general-purpose editor.

::: {.callout-note}
You do not have to give up VS Code. Positron keeps the extensions and settings you install separate from your VS Code installation, and they will not interfere with it. You can use both editors side by side.
:::

::: {.callout-note}
Positron also ships with an interactive **Migrating from VS Code to Positron** walkthrough. Open it anytime from the Command Palette by searching for _Welcome: Open Walkthrough_ and selecting it. Treat it as a quick-reference companion to this tutorial.

> :::

Reviewer: Can we re-organize this so we don't have callout notes back to back?

Maybe just convert one to "important"?

Author: Certainly. Done.

# Bring your settings and keybindings across

> You have spent time tuning VS Code: your theme, your font size, your keybindings, your formatter. You do not have to redo any of that.

Reviewer: ```suggestion
You have spent time tuning VS Code, including your theme, your font size, your keybindings, your formatter. You do not have to redo any of that.
```

We recommend importing your settings so Positron starts out feeling like the editor you already configured.

::: {.callout-tip title="Try it now"}
Open the Command Palette with {{< kbd mac=Command-Shift-P win=Ctrl-Shift-P linux=Ctrl-Shift-P >}}, search for _Preferences: Import Settings_, and run it. Review the preview, then confirm.

**You will know it worked when** your familiar theme and settings appear in Positron.
:::

![To import VS Code settings to Positron, run _Preferences: Import Settings_ then click Accept.](images/tutorial-vscode-import-settings.png){fig-alt="A screenshot displaying the Command Palette with Preferences: Import Settings selected and a pop up window asking the user if they would like to save imported VS Code settings."}

# Install your extensions from a different marketplace

Extensions work in Positron much as they do in VS Code: open the Extensions view from the **Activity Bar**, or run _Extensions: Focus on Extensions View_ from the Command Palette, then search for Extensions by name or keyword.

> The one real difference is where the extensions come from. Positron does not use the Microsoft Visual Studio Marketplace. Microsoft reserves that marketplace for their proprietary VS Code builds. Instead, Positron installs extensions from [Posit Public Package Manager](https://p3m.dev), which serves the same catalog as [Open VSX](https://open-vsx.org/), the vendor-neutral, open source marketplace for VS Code. Here, you will find over 10,000 extensions, including those published by data science software companies such as SAS, Snowflake, and Databricks. The extensions you will not find are the ones built with proprietary Microsoft code and licensed only for use inside the VS Code product from Microsoft.

Reviewer: ```suggestion
The one real difference is where the extensions come from. Positron does not use the Microsoft Visual Studio Marketplace. Microsoft reserves that marketplace for their proprietary VS Code builds. Instead, Positron installs extensions from [Posit Public Package Manager](https://p3m.dev), which serves the same catalog as [Open VSX](https://open-vsx.org/), the vendor-neutral, open source marketplace for VS Code compatible extensions. Here, you will find over 10,000 extensions, including those published by data science software companies such as SAS, Snowflake, and Databricks. The extensions you will not find are the ones built with proprietary Microsoft code and licensed only for use inside the VS Code product from Microsoft.
```

> Here is some more good news: you do not need the Microsoft Python and Jupyter extensions you rely on in VS Code. Positron provides their functionality natively, with no setup. Out of the box you already have:

Reviewer: ```suggestion
Some more good news is that you do not need the Microsoft Python and Jupyter extensions you rely on in VS Code. Positron provides their functionality natively, with no setup. Out of the box you already have:
```

| In Positron | In VS Code |
|---|---|
| Native Python support: code completion, type checking, debugging, and more, built in | Requires Microsoft Python extension (`ms-python.python`)|
| Bundled Jupyter notebook support | Requires Jupyter extension |
| Bundled [Ruff](https://docs.astral.sh/ruff/) for linting and formatting | Requires you to install and configure a formatter and linter, like Black Formatter and Pylint | 
| Bundled [Quarto](https://quarto.org/) support | Requires a separate publishing workflow | 

::: {.callout-note}
Are you an R user? You no longer need the `vscode-r` extension. Positron provides deeper, native R support, the same way it does for Python.
:::

::: {.callout-tip title="Try it now"}
Open the Extensions view and search for an extension you like, such as `Rainbow CSV`. Click **Install**.

**You will know it worked when** the extension installs from Posit Public Package Manager, exactly as it does from the marketplace in VS Code.
:::

::: {.callout-note}

> If an extension you rely on is missing, it might not be published to Open VSX. You can ask its author to [publish it to Open VSX](https://github.com/open-vsx/publish-extensions/blob/master/docs/external_contribution_request.md). See [Add features with extensions](extensions.qmd) for the full picture.

Reviewer: ```suggestion
If an extension you rely on is missing, it might not be published to Open VSX. You can ask its author to [publish it to Open VSX](https://github.com/open-vsx/publish-extensions/blob/master/docs/external_contribution_request.md). See [Extensions](extensions.qmd) for the full picture.
```

Author: Added.

:::

# Meet the Console

> In the last tutorial you did everything inside notebook cells. That works well for the analysis you want to keep. But often you just want to poke at your data: check the shape of a data frame, test whether a line of code runs, or try an idea you will throw away.

Reviewer: ```suggestion
In the last tutorial you did everything inside notebook cells. That works well for the analysis you want to keep. But often you just want to poke at your data, check the shape of a data frame, test whether a line of code runs, or try an idea you will throw away.
```

> In VS Code you can open a Python read-eval-print loop (REPL) in the Terminal or spin up an interactive window for this. Positron gives you the **Console**: a persistent, interactive Python session that is a first-class part of the IDE.

Reviewer: ```suggestion
In VS Code you can open a Python read-eval-print loop (REPL) in the Terminal or spin up an interactive window for this. Positron gives you the **Console**, a persistent, interactive Python session that is a first-class part of the IDE.
```

> The Console is more than a terminal REPL. Because Positron runs the interpreter as a live session, the Console offers code completion, syntax highlighting, and, most importantly, it drives the Variables and Plots panes you will meet next. It also knows your data at runtime: it can complete your data frame's column names, not just your variable names.

Reviewer: ```suggestion
The Console is more than a terminal REPL. Because Positron runs the interpreter as a live session, the Console offers code completion, syntax highlighting, and, most importantly, it drives the Variables and Plots panes you will meet next. It also knows your data at runtime; it can complete your data frame's column names, not just your variable names.
```

::: {.callout-tip title="Try it now"}
Focus the Console (run _Focus Console_ from the Command Palette, or click the Console pane). Type:

```python
from plotnine.data import mpg
mpg.shape
```

**You will know it worked when** the Console prints the shape of the data frame, and `mpg` appears in the Variables pane, without you adding a cell to your notebook.

:::

![Running commands in the Console.](images/tutorial-vscode-console.png){fig-alt="A screenshot that shows code typed into the Console pane."}

## Check your code against another environment

> Positron lets you run several Console sessions at once, each on a different Python version or environment. That makes it simple to answer a common question: will the code in my cell also run somewhere else?

Reviewer: ```suggestion
Positron lets you run several Console sessions at once, each on a different Python version or environment. That makes it simple to answer a common question, such as whether the code in my cell can also run somewhere else?
```

> Say you want to confirm your analysis works on a different Python version, or in a coworker's environment. Open a second Console session on that interpreter and run the same code. Both sessions stay open, and the **Interpreter picker** in the top right shows which one is active.

Reviewer: ```suggestion
Say you want to confirm your analysis works on a different Python version, or in a coworker's environment. Open a second Console session for that additional interpreter and run the same code. Both sessions stay open, and the **Interpreter picker** in the top right shows which one is active.
```

Author: Added.

:::

# Explore data without writing throwaway code

The Console feeds a set of data science panes that VS Code does not provide natively. You met two of them in the last tutorial. Here they are together:

- The **Variables pane** lists every object in your active session, with its type and value. You saw `mpg` appear there a moment ago.
- The **Data Explorer** opens any data frame in a sortable, filterable grid with per-column summary statistics. Its filters are ephemeral, so you can explore without changing your code, then click **Convert to Code** when you want to keep a step.
- The **Plots pane** collects the plots your code creates and keeps a history you can page through. You can export a plot or jump back to the code that produced it. You can drive the Plots pane with the Console, which means you no longer need a notebook just to see a plot.
- The **Help pane** shows documentation for Python objects and packages as you work.

::: {.callout-tip title="Try it now"}
In the Console, open your data frame in the Data Explorer by running:

```python

> view mpg

Reviewer: ```suggestion
%view mpg
```

> view mpg

Reviewer: It would be more credible for folks if we present this as a "magic":
https://positron.posit.co/guide-python.html#magics

The screenshot will also need to be updated.

Author: I made the changes that you suggested, but I wanted to flag this for you in case it seems noteworthy: my Positron had automagics turned on "out-of-the-box," which prevented the `%view` syntax from working. I fixed it by running `%automagic 0`. I don't use (or know about) magics, so I'd be surprised if it was something I did.

<img width="1475" height="873" alt="Screenshot 2026-07-21 at 4 12 20 PM" src="https://github.com/user-attachments/assets/2dcb903b-01fe-4f3c-b7a0-4befcbb3b88c" />

Reviewer: I just confirmed in a totally fresh build of Positron that `%view` is the right one that works by default.

```

**You will know it worked when** `mpg` opens in a grid. Sort a column by clicking its header, then open the summary panel to see its distribution.

:::

> ![Run `view` to view a table in the Data Explorer.](images/tutorial-vscode-data-explorer.png){fig-alt="A screenshot displaying the mpg data set in the Data Explorer and the line `view mpg` in the Console pane."}

Reviewer: ```suggestion
![Run `%view` to view a table in the Data Explorer.](images/tutorial-vscode-data-explorer.png){fig-alt="A screenshot displaying the mpg data set in the Data Explorer and the line `view mpg` in the Console pane."}
```

You can also open the Data Explorer by clicking the {{< fa table-cells >}} next to `mpg` in the Variables pane.

# Your notebook, upgraded

You might have noticed the notebook in the last tutorial looked a little different from the one in VS Code. That is because Positron uses an enhanced notebook editor built for data science. It comes with notebook-aware AI assistance and is synchronized with the Variables, Plots, Help, History, and Data Explorer panes you just used.

If you prefer the editor you know from VS Code, Positron still includes it as the [Legacy Notebook Editor](legacy-notebook-editor.qmd), and you can switch between the two.

::: {.callout-tip title="Try it now"}
If you did not try the Positron Notebook editor in the last tutorial, try it now:

1. Run _Create: New Jupyter Notebook_
2. Experiment with adding cells and running code, like `from plotnine.data import mpg`
3. Watch the Variables, Plots, Help, History, and Data Explorer panes interact with the notebook
:::

The video below demonstrates how the Positron IDE responds to your notebook edits, which enhances your authoring experience.

> {{< video videos/tutorial-vscode-notebooks.mp4 aria-label="The entire IDE tracks and responds to the Notebook Editor." >}}

Reviewer: This video has a lot of squiggles/diagnostics in it. Do you know what they are triggering for? We'd want to show a video without those squiggles.

Author: Fixed now. I tracked it down to ruff being unhappy I used 
```python
from plotnine import *
```
instead of 
```python
from plotnine import aes, geom_jitter, geom_smooth, ggplot, labs
```

# Set up AI with your Copilot account

If you use GitHub Copilot in VS Code, you do not have to leave it behind. Positron includes Posit Assistant, an AI assistant designed for data science, and it can use your existing GitHub Copilot account as its model provider.

We recommend adopting Posit Assistant and pointing it at the Copilot subscription you already pay for.

::: {.callout-tip title="Try it now"}
Run _Authentication: Configure Language Model Providers_ from the Command Palette, select **GitHub Copilot**, and sign in through your browser.

**You will know it worked when** Posit Assistant is connected and you can start a chat.
:::

![You can configure Posit Assistant to work with many model providers, including Github Copilot.](images/tutorial-vscode-copilot.png){fig-alt="The Configure Language Model Providers window." width=600}

::: {.callout-note}
GitHub Copilot is a preview provider. For account requirements and other providers, see [Posit Assistant](assistant.qmd) and [Language Model Providers](assistant-providers.qmd).
:::

# Let Positron manage your Python environments

> In VS Code you often install a Python version yourself and then point the editor at it. Positron does more of this for you.

Reviewer: ```suggestion
In VS Code you often install a Python version yourself and then point the editor at it. Positron can do more of this for you.
```

Author: Added

::: {.callout-note}
When Positron detects a `pyproject.toml` or `requirements.txt` in a project with no environment yet, it offers to create a `.venv` and install the dependencies with uv. See [Python installations and environments](python-installations.qmd) for the full set of options and settings, such as [`python.interpreters.include`](positron://settings/python.interpreters.include).
:::

# Know the gaps

Migrating confidently means knowing where the edges are. For VS Code users, the main one is the marketplace:

- Extensions that contain proprietary Microsoft code are not available, because Microsoft licenses them only for its own VS Code product. Open source extensions, which cover the great majority of data science tooling, work in Positron. If a specific extension matters to your workflow, check whether it is published to [Open VSX](https://open-vsx.org/) before you rely on it.

> 

Reviewer: ```suggestion
Migrating confidently means knowing where the edges are. For VS Code users, the main one is the marketplace. Extensions that contain proprietary Microsoft code are not available, because Microsoft licenses them only for its own VS Code product. Open source extensions, which cover the great majority of data science tooling, work in Positron. If a specific extension matters to your workflow, check whether it is published to [Open VSX](https://open-vsx.org/) before you rely on it.

```
